src/cplscheme/MultiCouplingScheme.cpp

```diff
@@ -82,41 +82,47 @@ void MultiCouplingScheme::exchangeInitialData()
   PRECICE_ASSERT(isImplicitCouplingScheme(), "MultiCouplingScheme is always Implicit.");
 
   if (_isController) {
-    if (receivesInitializedData()) {
-      for (auto &receiveExchange : _receiveDataVector) {
-        receiveData(_m2ns[receiveExchange.first], receiveExchange.second);
-      }
-      notifyDataHasBeenReceived();
-    } else {
-      for (auto &receiveExchange : _receiveDataVector) {
-        initializeWithZeroInitialData(receiveExchange.second);
+    for (auto &[from, data] : _receiveDataVector) {
+      if (receivesInitializedDataFrom(from)) {
+        receiveData(_m2ns.at(from), data);
+      } else {
+        initializeWithZeroInitialData(data);
       }
     }
-    if (sendsInitializedData()) {
-      for (auto &sendExchange : _sendDataVector) {
-        sendData(_m2ns[sendExchange.first], sendExchange.second);
+    notifyDataHasBeenReceived();
+    for (auto &[to, data] : _sendDataVector) {
+      if (sendsInitializedDataTo(to)) {
+        sendData(_m2ns.at(to), data);
       }
     }
   } else {
-    if (sendsInitializedData()) {
-      for (auto &sendExchange : _sendDataVector) {
-        sendData(_m2ns[sendExchange.first], sendExchange.second);
+    for (auto &[to, data] : _sendDataVector) {
+      if (sendsInitializedDataTo(to)) {
+        sendData(_m2ns.at(to), data);
       }
     }
-    if (receivesInitializedData()) {
-      for (auto &receiveExchange : _receiveDataVector) {
-        receiveData(_m2ns[receiveExchange.first], receiveExchange.second);
-      }
-      notifyDataHasBeenReceived();
-    } else {
-      for (auto &receiveExchange : _receiveDataVector) {
-        initializeWithZeroInitialData(receiveExchange.second);
+    for (auto &[from, data] : _receiveDataVector) {
+      if (receivesInitializedDataFrom(from)) {
+        receiveData(_m2ns.at(from), data);
+      } else {
+        initializeWithZeroInitialData(data);
       }
     }
+    notifyDataHasBeenReceived();
```

Comment:
Does it make sense to explicitly call `notifyDataHasBeenReceived(from)` inside the `for`-loop and then do a `PRECICE_CHECK` at the end to make sure all the required data has been received? (no need to do this here. Just an idea.)

Reply:
I think this is a common theme of this PR. Some functionality should be moved from the `BaseCouplingScheme` to either the `BiCouplingScheme` for two participant cases, and `MultiCouplingScheme` for multi-participant cases.

A more general alternative would be to change everything in the `BaseCouplingScheme` to a multi-participant format and always pass the coupling partner.

Reply:
> A more general alternative would be to change everything in the `BaseCouplingScheme` to a multi-participant format and always pass the coupling partner.

You mean similar to https://github.com/precice/precice/pull/1432 ? At that point in time we decided against the "make everything a multi coupling scheme"-approach.